### ISSUE TITLE
Design and build grouped time-series cross validation generator

### ISSUE THREAD

**Issue by SanthoshBala** [HealthRex/CDSS#89]
To avoid overfitting, we use cross-validation to split training data into multiple training/validation data sets.

The two most common methods for generating these cross-validation splits (k-fold and shuffle) assume that each sample is independent and identically distributed (i.i.d.).

Most of our learning applications break the i.i.d. assumption in two ways: 
1. Because the data set is constructed from event logs in an EMR, the samples are correlated with time (samples near each other in time are more likely to have similar features and labels).
2. Because the data set is constructed from a set of patient records, the samples are correlated with patient_id (samples from the same patient are more likely to have similar features and labels).

Both issues can cause information about the validation set to leak into the training set, causing us to underestimate the generalization error of the model.

Time series partitioning allows us to mitigate (1) by ensuring that the training set only contains samples from the "past" and the test set only contains samples from the "future."
Grouping allows us to mitigate (2) by ensuring that any single patient either falls into the training set or the validation set.

However, our data is typically structured as a set of groups, where each group is a time-series (e.g. all the events leading up to a patient's death, or all of the lab test orders during their hospitalization).

How might we design a cross-validation strategy to mitigate _both_  (1) and (2)?
